File: 321__Create_Maximum_Number.py
class Solution(object):
    def maxNumber(self, nums1, nums2, k):
        """
        :type nums1: List[int]
        :type nums2: List[int]
        :type k: int
        :rtype: List[int]
        """
        result = []
        for i in range(0, k + 1):
            if i <= len(nums1) and k - i <= len(nums2):
                sort1 = self.findmax(nums1, i)
                sort2 = self.findmax(nums2, k - i)
                merged = self.merge(sort1, sort2)
                if self.larger(merged, result):
                    result = merged
        return result

    def findmax(self, num, k):
        queue = []
        for i in range(0, len(num)):
            while len(queue) > 0 and num[i] > queue[-1] and len(queue) + len(num) - i > k:
                queue.pop()
            if len(queue) < k:
                queue.append(num[i])
        return queue

    def merge(self, sort1, sort2):
        i = 0
        j = 0
        ans = []
        while i < len(sort1) and j < len(sort2):
            if self.larger(sort1[i:], sort2[j:]):
                ans.append(sort1[i])
                i += 1
            else:
                ans.append(sort2[j])
                j += 1
        if i == len(sort1):
            ans += sort2[j:]
        else:
            ans += sort1[i:]
        return ans

    def larger(self, merged, ans):
        for i in range(0, min(len(merged), len(ans))):
            if merged[i] < ans[i]:
                return False
            elif merged[i] > ans[i]:
                return True
        if len(merged) < len(ans):
            return False
        else:
            return True

File: Java/321__Create_Maximum_Number.java
class Solution {
    public int[] maxNumber(int[] nums1, int[] nums2, int k) {
        int m = nums1.length, n = nums2.length;
        int[] result = new int[k];
        for (int i = Math.max(0, k - n); i <= k && i <= m; i++) {
            int[] candidate = merge(maxArray(nums1, i), maxArray(nums2, k - i), k);
            if (greater(candidate, 0, result, 0))
                result = candidate;
        }
        return result;
    }

    private int[] merge(int[] nums1, int[] nums2, int k) {
        int[] result = new int[k];
        for (int i = 0, j = 0, r = 0; r < k; ++r)
            result[r] = greater(nums1, i, nums2, j) ? nums1[i++] : nums2[j++];
        return result;
    }

    private boolean greater(int[] nums1, int i, int[] nums2, int j) {
        while (i < nums1.length && j < nums2.length && nums1[i] == nums2[j]) {
            i++;
            j++;
        }
        return j == nums2.length || (i < nums1.length && nums1[i] > nums2[j]);
    }

    private int[] maxArray(int[] nums, int k) {
        int n = nums.length;
        int[] result = new int[k];
        int idx = 0;
        for (int i = 0; i < n; i++) {
            while (n - i + idx > k && idx > 0 && result[idx - 1] < nums[i])
                idx--;
            if (idx < k) result[idx++] = nums[i];
        }
        return result;
    }
}
